File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
k, n = map(int, input().split())
arr = [int(input()) for _ in range(k)]

start, end = 1, max(arr)

while end>=start:
    mid = (start + end) // 2
    cnt = 0
    for i in arr:
        cnt += i//mid

    if cnt>=n:
        start = mid+1
    elif cnt<n:
        end = mid-1

print(end)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int k = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());
        int[] arr = new int[k];
        long end = 0;
        for (int i = 0; i<k; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            if (end < arr[i]) {
                end = arr[i];
            }
        }
        end++;
        long start = 0;
        long mid = 0;
        while (start<end) {
            mid = (start+end)/2;
            long count = 0;
            for (int i = 0; i<k; i++) {
                count += (arr[i] / mid);
            }
            if (count<n) {
                end = mid;
            } else {
                start = mid+1;
            }
        }
        System.out.println(start-1);
    }
}